src/Hl7.Fhir.Core/Serialization/ComplexTypeWriter.cs

```diff
@@ -47,120 +49,104 @@ internal void Serialize(ClassMapping mapping, object instance, Rest.SummaryType
 
             _writer.WriteStartComplexContent();
 
-            // Emit members that need xml /attributes/ first (to facilitate stream writer API)
-            foreach (var prop in mapping.PropertyMappings.Where(pm => pm.SerializationHint == XmlSerializationHint.Attribute))
+            // Emit members that need xml attributes / first (to facilitate stream writer API)
+            // attributes first (xml) and order maintained for the rest
+            var propertiesToWrite = mapping.PropertyMappings
+                .Where(property => property.SerializationHint == XmlSerializationHint.Attribute)
+                .Concat(mapping.PropertyMappings.Where(property => property.SerializationHint != XmlSerializationHint.Attribute));
+
+            if (summary == Rest.SummaryType.True)
             {
-                writeProperty(mapping, instance, summary, mode, prop);
+                propertiesToWrite = propertiesToWrite.Where(property => property.InSummary);
             }
-
-            // Then emit the rest
-            foreach (var prop in mapping.PropertyMappings.Where(pm => pm.SerializationHint != XmlSerializationHint.Attribute))
+            else if (summary == Rest.SummaryType.Text)
             {
-                writeProperty(mapping, instance, summary, mode, prop);
+                propertiesToWrite = propertiesToWrite.Where(property =>
+                       summaryTextProperties.Contains(property.Name.ToLower())
```

Comment:
The original code also checked the type of the property, to try to avoid serializing properties that have accidentally the same name as the intended properties.  It's still not bullet-proof, since we would need the parent path to be sure, but it helps.